Fix control flow with multiple exits on single-pass BBQ JIT

ddegazio opened the pull request:
#### bc648d25c5aa6be1fffc4794700336f357742a45
<pre>
Fix control flow with multiple exits on single-pass BBQ JIT
<a href="https://bugs.webkit.org/show_bug.cgi?id=253298">https://bugs.webkit.org/show_bug.cgi?id=253298</a>
rdar://106182518

Reviewed by Yusuke Suzuki.

Fixes br_table, func, switch WASM spec tests when using single-pass
BBQ JIT on Intel. Specifically, refactors BBQ JIT&apos;s control flow data
to support passing values to multiple distinct exit branches per block.

* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::ControlData::flushAtBlockBoundary):
(JSC::Wasm::BBQJIT::ControlData::addExit):
(JSC::Wasm::BBQJIT::ControlData::finalizeBlock):
(JSC::Wasm::BBQJIT::ControlData::flushAndSingleExit):
(JSC::Wasm::BBQJIT::ControlData::targetLocations const):
(JSC::Wasm::BBQJIT::addBlock):
(JSC::Wasm::BBQJIT::addLoop):
(JSC::Wasm::BBQJIT::addIf):
(JSC::Wasm::BBQJIT::addElse):
(JSC::Wasm::BBQJIT::addTry):
(JSC::Wasm::BBQJIT::addCatch):
(JSC::Wasm::BBQJIT::addCatchAll):
(JSC::Wasm::BBQJIT::addBranch):
(JSC::Wasm::BBQJIT::addSwitch):
(JSC::Wasm::BBQJIT::addEndToUnreachable):
(JSC::Wasm::BBQJIT::ControlData::endBlock): Deleted.

Canonical link: <a href="https://commits.webkit.org/261122@main">https://commits.webkit.org/261122@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/484bfc568318e07a2d7927d3631012146680baa9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110684 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19765 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/43247 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/2022 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/119519 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21174 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10877 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/1545 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116425 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15770 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/43247 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102985 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97731 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/43247 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/44100 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/99345 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/12400 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/43247 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/85993 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/83/builds/10473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12949 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8916 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/100411 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18322 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/43247 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/31322 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/14887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/108439 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4200 "Built successfully and passed tests") | | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/26731 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -6541,41 +6568,54 @@ class BBQJIT {
             int targetIndex = condition.asI32();
-            if (targetIndex >= 0 && targetIndex < static_cast<int>(targets.size()))
+            if (targetIndex >= 0 && targetIndex < static_cast<int>(targets.size())) {
+                currentControlData().flushAndSingleExit(*this, *targets[targetIndex], m_parser->expressionStack(), false, true);
```
How about passing `results` instead of `m_parser->expressionStack()`? They are the same.

ddegazio replied:
Looks like this is unnecessary...I thought this was the cause of some debug assertion failures, but it turns out that was a separate issue. This is resolved in the latest push.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -6541,41 +6568,54 @@ class BBQJIT {
                 targets[targetIndex]->addBranch(m_jit.jump());
-            else
+            } else {
+                currentControlData().flushAndSingleExit(*this, defaultTarget, m_parser->expressionStack(), false, true);
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -6587,11 +6627,15 @@ class BBQJIT {
             unsigned index = binarySwitch.caseIndex();
             ASSERT_UNUSED(value, value == index);
             ASSERT(index < targets.size());
+            currentControlData().addExit(*this, targets[index]->targetLocations(), m_parser->expressionStack());
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -6587,11 +6627,15 @@ class BBQJIT {
 
-        defaultTarget.addBranch(binarySwitch.fallThrough());
+        binarySwitch.fallThrough().link(&m_jit);
+        currentControlData().addExit(*this, defaultTarget.targetLocations(), m_parser->expressionStack());
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -6587,11 +6627,15 @@ class BBQJIT {
+        currentControlData().addExit(*this, defaultTarget.targetLocations(), m_parser->expressionStack());
+        defaultTarget.addBranch(m_jit.jump());
 
+        currentControlData().finalizeBlock(*this, defaultTarget.targetLocations().size(), m_parser->expressionStack(), false);
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -6541,41 +6568,54 @@ class BBQJIT {
 
     PartialResult WARN_UNUSED_RETURN addSwitch(Value condition, const Vector<ControlData*>& targets, ControlData& defaultTarget, Stack& results)
     {
+        UNUSED_PARAM(results);
```
If we use results, then we do not need this.

Constellation reviewed (approved):
r=me

ddegazio committed:
Fix control flow with multiple exits on single-pass BBQ JIT
https://bugs.webkit.org/show_bug.cgi?id=253298
rdar://106182518

Reviewed by Yusuke Suzuki.

Fixes br_table, func, switch WASM spec tests when using single-pass
BBQ JIT on Intel. Specifically, refactors BBQ JIT's control flow data
to support passing values to multiple distinct exit branches per block.

* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::ControlData::flushAtBlockBoundary):
(JSC::Wasm::BBQJIT::ControlData::addExit):
(JSC::Wasm::BBQJIT::ControlData::finalizeBlock):
(JSC::Wasm::BBQJIT::ControlData::flushAndSingleExit):
(JSC::Wasm::BBQJIT::ControlData::targetLocations const):
(JSC::Wasm::BBQJIT::addBlock):
(JSC::Wasm::BBQJIT::addLoop):
(JSC::Wasm::BBQJIT::addIf):
(JSC::Wasm::BBQJIT::addElse):
(JSC::Wasm::BBQJIT::addTry):
(JSC::Wasm::BBQJIT::addCatch):
(JSC::Wasm::BBQJIT::addCatchAll):
(JSC::Wasm::BBQJIT::addBranch):
(JSC::Wasm::BBQJIT::addSwitch):
(JSC::Wasm::BBQJIT::addEndToUnreachable):
(JSC::Wasm::BBQJIT::ControlData::endBlock): Deleted.

Canonical link: https://commits.webkit.org/261122@main
